# Multifasta_analysis_tool

This tool is used to filter and manipulate of multifasta files. Also you can extract genes from the files !

## Installation

Make sure that you have python 3.7 or above and all the dependacies (Biopython, matplotlib, pandas , numpy , difflib, re )

The tool was written to be used on Linux command line or any python IDE !

## Usage
For your own comfort, make sure you have the files in the same directory as the tool!

```python
python3 Multifasta_analysis_tool.py

```
then just answer questions !

## This tool can be used if..
1- you want extract a gene using its start/end postion in aligned_file.afa (more accurate) or multi/fasta file

2- you want to exlude sequences in a multifasta file using a pattern in the sequneces (ex:NNN,XX)

3- you want to  print all  ID headers in your multifasta

4- you want extract a gene from fasta/multifasta file using it's upstream and downstream sequence

5- you want to translat DNA fasta file on the its 1 frame

6- you want to efetch data from NCBI or you want to merge your multifasta files in accurate and fast way 

7- you want to know GC content and N bases content of your DNA seq

8- you want to align muscle or mafft

9- you want to draw a tree.dnd using python matplotlib

10- you want to extract the longest conserved sequnece and you want to call variants between your  aligned genomes , genes or proteins from a clustal_file.aln and the output is just fasta for your conserved sequnec and xlsx file for variants with a graph where is a map for the mutation 

## Contributing
Pull requests are very welcome. For major changes, please open an issue first to discuss what you would like to change.

The tool is still under development !

Contact me directly on email : drahmedsherbini@yahoo.com
## License
Please, cite my page if this tool was useful for your work

